lib/bn/bn128_arithmetic.ex

```diff
@@ -86,72 +110,113 @@ defmodule BN.BN128Arithmetic do
+  @spec double(point()) :: point()
+  def double({x, y} = {%FQ{}, %FQ{}}) do
+    if y.value == 0 do
+      {FQ.new(0), FQ.new(0)}
+    else
+      double_y = FQ.mult(y, 2)
+
+      lambda =
+        x
+        |> FQ.pow(2)
+        |> FQ.mult(3)
+        |> FQ.divide(double_y)
+
+      double_x = FQ.mult(x, 2)
+
+      new_x = lambda |> FQ.pow(2) |> FQ.sub(double_x)
+
+      new_y =
+        x
+        |> FQ.sub(new_x)
+        |> FQ.mult(lambda)
+        |> FQ.sub(y)
+
+      {new_x, new_y}
+    end
+  end
 
-      result
+  def double({x, y} = {%FQP{}, %FQP{}}) do
+    if y == FQ12.zero() do
+      {FQ12.zero(), FQ12.zero()}
     else
-      double_y = IntegerModP.mult(point.y, 2)
+      double_y = FQ12.mult(y, 2)
 
       lambda =
-        point.x
-        |> IntegerModP.pow(2)
-        |> IntegerModP.mult(3)
-        |> IntegerModP.div(double_y)
+        x
+        |> FQ12.pow(2)
+        |> FQ12.mult(3)
+        |> FQ12.divide(double_y)
 
-      double_x = IntegerModP.mult(point.x, 2)
+      double_x = FQ12.mult(x, 2)
 
-      x = lambda |> IntegerModP.pow(2) |> IntegerModP.sub(double_x)
+      new_x = lambda |> FQ12.pow(2) |> FQ12.sub(double_x)
 
-      y =
-        point.x
-        |> IntegerModP.sub(x)
-        |> IntegerModP.mult(lambda)
-        |> IntegerModP.sub(point.y)
-
-      %Point{
-        x: x,
-        y: y,
-        modulus: point.modulus
-      }
+      new_y =
+        x
+        |> FQ12.sub(new_x)
+        |> FQ12.mult(lambda)
+        |> FQ12.sub(y)
+
+      {new_x, new_y}
     end
   end
 
-  @spec calculate_points_addition(Point.t(), Point.t()) :: Point.t()
-  defp calculate_points_addition(point1, point2) do
-    if point1.x == point2.x do
-      {:ok, result} = Point.new(0, 0, modulus: point1.modulus)
+  @spec calculate_points_addition(point(), point()) :: point()
+  defp calculate_points_addition({x1, y1} = {%FQ{}, %FQ{}}, {x2, y2}) do
+    if x1 == x2 do
+      {FQ.new(0), FQ.new(0)}
+    else
+      y_remainder = FQ.sub(y2, y1)
+      x_remainder = FQ.sub(x2, x1)
+      lambda = FQ.divide(y_remainder, x_remainder)
+
+      x =
+        lambda
+        |> FQ.pow(2)
+        |> FQ.sub(x1)
+        |> FQ.sub(x2)
+
+      y =
+        x1
+        |> FQ.sub(x)
+        |> FQ.mult(lambda)
+        |> FQ.sub(y1)
+
+      {x, y}
+    end
+  end
 
-      result
+  defp calculate_points_addition({x1, y1} = {%FQP{}, %FQP{}}, {x2, y2}) do
+    if x1 == x2 do
+      {FQ12.zero(), FQ12.zero()}
     else
-      y_remainder = IntegerModP.sub(point2.y, point1.y)
-      x_remainder = IntegerModP.sub(point2.x, point1.x)
-      lambda = IntegerModP.div(y_remainder, x_remainder)
+      y_remainder = FQ12.sub(y2, y1)
+      x_remainder = FQ12.sub(x2, x1)
+      lambda = FQ12.divide(y_remainder, x_remainder)
 
       x =
         lambda
-        |> IntegerModP.pow(2)
-        |> IntegerModP.sub(point1.x)
-        |> IntegerModP.sub(point2.x)
+        |> FQ12.pow(2)
+        |> FQ12.sub(x1)
+        |> FQ12.sub(x2)
```

Comment:
Seems like we have a lot of these `lambda`s that are all similar. Would it make sense to extract functions out of them?